libfreerdp/core/aad.c

```diff
@@ -0,0 +1,1148 @@
+	*result = str;
+	rc = str != NULL;
+
+fail:
+	cJSON_Delete(prop);
+	return rc;
+}
+
+static BOOL json_get_string_alloc(wLog* wlog, cJSON* json, const char* key, char** result)
+{
+	const char* str = NULL;
+	if (!json_get_const_string(wlog, json, key, &str))
+		return FALSE;
+	free(*result);
+	*result = _strdup(str);
+	if (!*result)
+		WLog_Print(wlog, WLOG_ERROR, "[json] object for key '%s' strdup is NULL", key);
+	return *result != NULL;
+}
+
+static BIO* aad_connect_https(rdpAad* aad, SSL_CTX* ssl_ctx)
+{
+	WINPR_ASSERT(aad);
+	WINPR_ASSERT(ssl_ctx);
+
+	const int vprc = SSL_CTX_set_default_verify_paths(ssl_ctx);
+	if (vprc != 1)
+	{
+		WLog_Print(aad->log, WLOG_ERROR, "Error setting verify paths");
+		return NULL;
+	}
+	const long mrc = SSL_CTX_set_mode(ssl_ctx, SSL_MODE_AUTO_RETRY);
+
+	BIO* bio = BIO_new_ssl_connect(ssl_ctx);
+	if (!bio)
+	{
+		WLog_Print(aad->log, WLOG_ERROR, "Error setting up connection");
+		return NULL;
+	}
+	const long chrc = BIO_set_conn_hostname(bio, auth_server);
+	if (chrc != 1)
+	{
+		WLog_Print(aad->log, WLOG_ERROR, "Error setting BIO hostname");
+		BIO_free(bio);
+		return NULL;
+	}
+	const long cprc = BIO_set_conn_port(bio, "https");
+	if (cprc != 1)
+	{
+		WLog_Print(aad->log, WLOG_ERROR, "Error setting BIO port");
+		BIO_free(bio);
+		return NULL;
+	}
+	return bio;
+}
+
+static BOOL aad_logging_bio_write(rdpAad* aad, BIO* bio, const char* str)
+{
+	WINPR_ASSERT(aad);
+	WINPR_ASSERT(bio);
+	WINPR_ASSERT(str);
+
+	const size_t size = strlen(str);
+	if (size > INT_MAX)
+		return FALSE;
+	ERR_clear_error();
+	if (BIO_write(bio, str, (int)size) < 0)
+	{
+		ERR_print_errors_cb(print_error, aad->log);
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static char* aad_read_response(rdpAad* aad, BIO* bio, size_t* plen, const char* what)
+{
+	WINPR_ASSERT(plen);
+
+	long status_code;
+	char* buffer = NULL;
+	size_t length = 0;
+
+	*plen = 0;
+	if (!read_http_message(aad, bio, &status_code, &buffer, &length))
+	{
+		WLog_Print(aad->log, WLOG_ERROR, "Unable to read %s HTTP response", what);
+		return NULL;
+	}
+	WLog_Print(aad->log, WLOG_DEBUG, "%s HTTP response: %s", buffer);
+
+	if (status_code != 200)
+	{
+		WLog_Print(aad->log, WLOG_ERROR, "%s HTTP status code: %li", status_code);
+		free(buffer);
+		return NULL;
+	}
+	*plen = length;
+	return buffer;
+}
+
+static BOOL aad_read_and_extract_token_from_json(rdpAad* aad, BIO* bio)
+{
+	BOOL rc = FALSE;
+	size_t blen = 0;
+	char* buffer = aad_read_response(aad, bio, &blen, "access token");
+	if (!buffer)
+		return FALSE;
+
+	cJSON* json = cJSON_ParseWithLength(buffer, blen);
+	if (!json)
+	{
+		WLog_Print(aad->log, WLOG_ERROR, "Failed to parse JSON response");
+		goto fail;
+	}
+
+	if (!json_get_string_alloc(aad->log, json, "access_token", &aad->access_token))
+	{
+		WLog_Print(aad->log, WLOG_ERROR,
+		           "Could not find \"access_token\" property in JSON response");
+		goto fail;
+	}
+
+	rc = TRUE;
+fail:
+	free(buffer);
+	cJSON_free(json);
```

Comment:
`cJSON_free()` only frees the object pointed to itself. Better use `cJSON_Delete()`